### PR TITLE
Support the user preference for trailing-new-lines when moving types.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/17171

### CODE REVIEW

File: src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs [hunk 1]
```diff
@@ -7,22 +7,31 @@ namespace Microsoft.CodeAnalysis.Formatting
...
 
-        public static PerLanguageOption<string> NewLine { get; } = new PerLanguageOption<string>(nameof(FormattingOptions), nameof(NewLine), defaultValue: "\r\n",
-            storageLocations: new EditorConfigStorageLocation("end_of_line", ParseEditorConfigEndOfLine));
+        public static PerLanguageOption<string> NewLine { get; } =
+            new PerLanguageOption<string>(nameof(FormattingOptions), nameof(NewLine), defaultValue: "\r\n",
+                storageLocations: new EditorConfigStorageLocation("end_of_line", ParseEditorConfigEndOfLine));
+
+        internal static PerLanguageOption<bool> InsertFinalNewLine { get; } =
+            new PerLanguageOption<bool>(nameof(FormattingOptions), nameof(InsertFinalNewLine), defaultValue: false,
+                storageLocations: new EditorConfigStorageLocation("insert_final_newline", s => s == "true"));
```
Comment (jmarolf) [inline src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs]: >, s => s == "true" [](start = 88, length = 18)

Delete this, the default parser will handle it.

File: src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs [hunk 2]
```diff
@@ -7,22 +7,31 @@ namespace Microsoft.CodeAnalysis.Formatting
...
+        public static PerLanguageOption<IndentStyle> SmartIndent { get; } =
+            new PerLanguageOption<IndentStyle>(nameof(FormattingOptions), nameof(SmartIndent), defaultValue: IndentStyle.Smart);
 
-        public static PerLanguageOption<string> NewLine { get; } = new PerLanguageOption<string>(nameof(FormattingOptions), nameof(NewLine), defaultValue: "\r\n",
-            storageLocations: new EditorConfigStorageLocation("end_of_line", ParseEditorConfigEndOfLine));
+        public static PerLanguageOption<string> NewLine { get; } =
+            new PerLanguageOption<string>(nameof(FormattingOptions), nameof(NewLine), defaultValue: "\r\n",
+                storageLocations: new EditorConfigStorageLocation("end_of_line", ParseEditorConfigEndOfLine));
+
+        internal static PerLanguageOption<bool> InsertFinalNewLine { get; } =
```
Comment (jmarolf) [inline src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs]: >PerLanguageOption [](start = 24, length = 17)

Why PerLanguageOption?

**Comment by jmarolf** [comment]
👎

File: src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs]: Cool!

File: src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs]: i was following end_of_line.  should i not be using that?

**Comment by CyrusNajmabadi** [comment]
@jmarolf Are you ok with this now?

**Comment by CyrusNajmabadi** [comment]
retest windows_debug_unit64_prtest  please

File: src/Compilers/Core/Portable/Collections/BitVector.cs [hunk 3]
```diff
@@ -70,13 +70,7 @@ private static int WordsForCapacity(int capacity)
...
         }
 
-        public int Capacity
-        {
-            get
-            {
-                return _capacity;
-            }
-        }
+        public int Capacity => _capacity;
```
Comment (Pilchie) [inline src/Compilers/Core/Portable/Collections/BitVector.cs]: You changed a bunch of compiler files without talking to the compiler team.

**Pull request merged**